### ISSUE TITLE
AnyUrl: TypeError: object of type 'AnyUrl' has no len()

### ISSUE THREAD

**Issue by Kharianne** [pydantic/pydantic#11092]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Hello,

I have noticed that when using **AnyUrl** with **Field** **max_length** it returns:
```
Traceback (most recent call last):
  File "/Users/XXXX/XXX/XXXXX/.venv/lib/python3.12/site-packages/pydantic/_internal/_validators.py", line 324, in max_length_validator
    if len(x) > max_length:
       ^^^^^^
TypeError: object of type 'AnyUrl' has no len()
```

I'm aware that more Pydantic way would be to use `UrlConstrains` with max_length. However before introducing _BaseUrl and its custom schema (so in **Pydantic 2.9.2**), it was working and I notice that had been used in some libraries. 

I tried if simple adding a magic method __len__ to _BaseUrl would solve the issue and it worked. However I'm not sure if this would be the best solution (quite new to internals of Pydantic).
```python
def __len__(self) -> int:
    return len(str(self._url))
```

So to sum up, code below worked without an issue in Pydantic 2.9.2, however in Pydantic 2.10.x the same code raises TypeError.

### Example Code

```Python
from pydantic import AnyUrl, Field, BaseModel


class SomeClass(BaseModel):
    url: AnyUrl = Field(max_length=100)


SomeClass(url="https://example.com")
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.10.3
pydantic-core version: 2.27.1
pydantic-core build: profile=release pgo=false
install path: /Users/XXXXX/XXXX/XXXXXX/.venv/lib/python3.12/site-packages/pydantic
python version: 3.12.7 (main, Oct  1 2024, 02:05:46) [Clang 15.0.0 (clang-1500.3.9.4)]
platform: macOS-14.6.1-arm64-arm-64bit
related packages: typing_extensions-4.12.2
commit: unknown
```

**Comment by sydney-runkle** [comment]
Yep, I think a `__len__` addition would be great here. PRs welcome. We can include this in v2.10.4. Thanks!

**Comment by Kharianne** [comment]
@sydney-runkle I can make a PR later today or tomorrow morning :)
